syntax = "proto3";

package myproject.fakeapp;

import "google/protobuf/empty.proto";
import "google/protobuf/struct.proto";

service BasicController {
    rpc BasicList(BasicProtoListChildListRequest) returns (BasicProtoListChildListResponse) {}
    rpc Create(BasicServiceRequest) returns (BasicServiceResponse) {}
    rpc FetchDataForUser(BasicFetchDataForUserRequest) returns (BasicServiceResponse) {}
    rpc FetchTranslatedKey(google.protobuf.Empty) returns (BasicFetchTranslatedKeyResponse) {}
    rpc GetMultiple(google.protobuf.Empty) returns (BasicServiceListResponse) {}
    rpc ListIds(google.protobuf.Empty) returns (BasicListIdsResponse) {}
    rpc ListName(google.protobuf.Empty) returns (BasicListNameResponse) {}
    rpc MixParam(CustomMixParamForListRequest) returns (BasicMixParamListResponse) {}
    rpc MixParamWithSerializer(BasicParamWithSerializerListRequest) returns (BasicMixParamWithSerializerListResponse) {}
    rpc MyMethod(CustomNameForRequest) returns (CustomNameForResponse) {}
    rpc TestBaseProtoSerializer(BaseProtoExampleRequest) returns (BaseProtoExampleListResponse) {}
    rpc TestEmptyMethod(google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc TestNoMetaSerializer(NoMetaRequest) returns (BasicTestNoMetaSerializerResponse) {}
}

service DefaultValueController {
    rpc Create(DefaultValueRequest) returns (DefaultValueResponse) {}
    rpc Destroy(DefaultValueDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(DefaultValueListRequest) returns (DefaultValueListResponse) {}
    rpc PartialUpdate(DefaultValuePartialUpdateRequest) returns (DefaultValueResponse) {}
    rpc Retrieve(DefaultValueRetrieveRequest) returns (DefaultValueResponse) {}
    rpc Update(DefaultValueRequest) returns (DefaultValueResponse) {}
}

service ExceptionController {
    rpc APIException(google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc GRPCException(google.protobuf.Empty) returns (google.protobuf.Empty) {}
    rpc StreamRaiseException(google.protobuf.Empty) returns (stream ExceptionStreamRaiseExceptionResponse) {}
    rpc UnaryRaiseException(google.protobuf.Empty) returns (google.protobuf.Empty) {}
}

service ForeignModelController {
    rpc List(ForeignModelListRequest) returns (ForeignModelListResponse) {}
    rpc Retrieve(ForeignModelRetrieveCustomRetrieveRequest) returns (ForeignModelRetrieveCustomResponse) {}
}

service ImportStructEvenInArrayModelController {
    rpc Create(ImportStructEvenInArrayModelRequest) returns (ImportStructEvenInArrayModelResponse) {}
}

service RecursiveTestModelController {
    rpc Create(RecursiveTestModelRequest) returns (RecursiveTestModelResponse) {}
    rpc Destroy(RecursiveTestModelDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(RecursiveTestModelListRequest) returns (RecursiveTestModelListResponse) {}
    rpc PartialUpdate(RecursiveTestModelPartialUpdateRequest) returns (RecursiveTestModelResponse) {}
    rpc Retrieve(RecursiveTestModelRetrieveRequest) returns (RecursiveTestModelResponse) {}
    rpc Update(RecursiveTestModelRequest) returns (RecursiveTestModelResponse) {}
}

service RelatedFieldModelController {
    rpc Create(RelatedFieldModelRequest) returns (RelatedFieldModelResponse) {}
    rpc Destroy(RelatedFieldModelDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(RelatedFieldModelListRequest) returns (RelatedFieldModelListResponse) {}
    rpc PartialUpdate(RelatedFieldModelPartialUpdateRequest) returns (RelatedFieldModelResponse) {}
    rpc Retrieve(RelatedFieldModelRetrieveRequest) returns (RelatedFieldModelResponse) {}
    rpc Update(RelatedFieldModelRequest) returns (RelatedFieldModelResponse) {}
}

service SimpleRelatedFieldModelController {
    rpc Create(SimpleRelatedFieldModelRequest) returns (SimpleRelatedFieldModelResponse) {}
    rpc Destroy(SimpleRelatedFieldModelDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(SimpleRelatedFieldModelListRequest) returns (SimpleRelatedFieldModelListResponse) {}
    rpc PartialUpdate(SimpleRelatedFieldModelPartialUpdateRequest) returns (SimpleRelatedFieldModelResponse) {}
    rpc Retrieve(SimpleRelatedFieldModelRetrieveRequest) returns (SimpleRelatedFieldModelResponse) {}
    rpc Update(SimpleRelatedFieldModelRequest) returns (SimpleRelatedFieldModelResponse) {}
}

service SpecialFieldsModelController {
    rpc Create(SpecialFieldsModelRequest) returns (SpecialFieldsModelResponse) {}
    rpc Destroy(SpecialFieldsModelDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(SpecialFieldsModelListRequest) returns (SpecialFieldsModelListResponse) {}
    rpc PartialUpdate(SpecialFieldsModelPartialUpdateRequest) returns (SpecialFieldsModelResponse) {}
    rpc Retrieve(SpecialFieldsModelRetrieveRequest) returns (CustomRetrieveResponseSpecialFieldsModelResponse) {}
    rpc Update(SpecialFieldsModelRequest) returns (SpecialFieldsModelResponse) {}
}

service StreamInController {
    rpc StreamIn(stream StreamInStreamInRequest) returns (StreamInStreamInListResponse) {}
    rpc StreamToStream(stream StreamInStreamToStreamRequest) returns (stream StreamInStreamToStreamResponse) {}
    rpc StreamToStreamReadWrite(stream StreamInStreamToStreamRequest) returns (stream StreamInStreamToStreamRequest) {}
}

service SyncUnitTestModelController {
    rpc Create(UnitTestModelRequest) returns (UnitTestModelResponse) {}
    rpc Destroy(UnitTestModelDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(UnitTestModelListRequest) returns (UnitTestModelListResponse) {}
    rpc ListWithExtraArgs(SyncUnitTestModelListWithExtraArgsRequest) returns (UnitTestModelListExtraArgsResponse) {}
    rpc PartialUpdate(UnitTestModelPartialUpdateRequest) returns (UnitTestModelResponse) {}
    rpc Retrieve(UnitTestModelRetrieveRequest) returns (UnitTestModelResponse) {}
    rpc Stream(UnitTestModelStreamRequest) returns (stream UnitTestModelResponse) {}
    rpc Update(UnitTestModelRequest) returns (UnitTestModelResponse) {}
}

service UnitTestModelController {
    rpc Create(UnitTestModelRequest) returns (UnitTestModelResponse) {}
    rpc Destroy(UnitTestModelDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(UnitTestModelListRequest) returns (UnitTestModelListResponse) {}
    rpc ListWithExtraArgs(UnitTestModelListWithExtraArgsRequest) returns (UnitTestModelListExtraArgsResponse) {}
    rpc PartialUpdate(UnitTestModelPartialUpdateRequest) returns (UnitTestModelResponse) {}
    rpc Retrieve(UnitTestModelRetrieveRequest) returns (UnitTestModelResponse) {}
    rpc Stream(UnitTestModelStreamRequest) returns (stream UnitTestModelResponse) {}
    rpc Update(UnitTestModelRequest) returns (UnitTestModelResponse) {}
}

service UnitTestModelWithCacheController {
    rpc Create(UnitTestModelWithCacheRequest) returns (UnitTestModelWithCacheResponse) {}
    rpc Destroy(UnitTestModelWithCacheDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithAutoCacheCleanOnSaveAndDelete(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithAutoCacheCleanOnSaveAndDeleteRedis(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithPossibilityMaxAge(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithStructFilter(UnitTestModelWithCacheListWithStructFilterRequest) returns (UnitTestModelWithCacheListResponse) {}
    rpc PartialUpdate(UnitTestModelWithCachePartialUpdateRequest) returns (UnitTestModelWithCacheResponse) {}
    rpc Retrieve(UnitTestModelWithCacheRetrieveRequest) returns (UnitTestModelWithCacheResponse) {}
    rpc Stream(UnitTestModelWithCacheStreamRequest) returns (stream UnitTestModelWithCacheResponse) {}
    rpc Update(UnitTestModelWithCacheRequest) returns (UnitTestModelWithCacheResponse) {}
}

service UnitTestModelWithCacheInheritController {
    rpc Create(UnitTestModelWithCacheRequest) returns (UnitTestModelWithCacheResponse) {}
    rpc Destroy(UnitTestModelWithCacheDestroyRequest) returns (google.protobuf.Empty) {}
    rpc List(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithAutoCacheCleanOnSaveAndDelete(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithAutoCacheCleanOnSaveAndDeleteRedis(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithPossibilityMaxAge(google.protobuf.Empty) returns (UnitTestModelWithCacheListResponse) {}
    rpc ListWithStructFilter(UnitTestModelWithCacheInheritListWithStructFilterRequest) returns (UnitTestModelWithCacheListResponse) {}
    rpc PartialUpdate(UnitTestModelWithCachePartialUpdateRequest) returns (UnitTestModelWithCacheResponse) {}
    rpc Retrieve(UnitTestModelWithCacheRetrieveRequest) returns (UnitTestModelWithCacheResponse) {}
    rpc Stream(UnitTestModelWithCacheStreamRequest) returns (stream UnitTestModelWithCacheResponse) {}
    rpc Update(UnitTestModelWithCacheRequest) returns (UnitTestModelWithCacheResponse) {}
}

service UnitTestModelWithStructFilterController {
    rpc Create(UnitTestModelWithStructFilterRequest) returns (UnitTestModelWithStructFilterResponse) {}
    rpc Destroy(UnitTestModelWithStructFilterDestroyRequest) returns (google.protobuf.Empty) {}
    rpc EmptyWithFilter(UnitTestModelWithStructFilterEmptyWithFilterRequest) returns (google.protobuf.Empty) {}
    rpc List(UnitTestModelWithStructFilterListRequest) returns (UnitTestModelWithStructFilterListResponse) {}
    rpc PartialUpdate(UnitTestModelWithStructFilterPartialUpdateRequest) returns (UnitTestModelWithStructFilterResponse) {}
    rpc Retrieve(UnitTestModelWithStructFilterRetrieveRequest) returns (UnitTestModelWithStructFilterResponse) {}
    rpc Stream(UnitTestModelWithStructFilterStreamRequest) returns (stream UnitTestModelWithStructFilterResponse) {}
    rpc Update(UnitTestModelWithStructFilterRequest) returns (UnitTestModelWithStructFilterResponse) {}
}

message BaseProtoExampleListResponse {
    repeated BaseProtoExampleResponse results = 1;
    int32 count = 2;
}

message BaseProtoExampleRequest {
    string uuid = 1;
    int32 number_of_elements = 2;
    bool is_archived = 3;
}

message BaseProtoExampleResponse {
    string uuid = 1;
    int32 number_of_elements = 2;
    bool is_archived = 3;
}

message BasicFetchDataForUserRequest {
    string user_name = 1;
}

message BasicFetchTranslatedKeyResponse {
    string text = 1;
}

message BasicListIdsResponse {
    repeated int32 ids = 1;
}

message BasicListNameResponse {
    repeated string name = 1;
}

message BasicMixParamListResponse {
    repeated BasicMixParamResponse results = 1;
    int32 count = 2;
}

message BasicMixParamResponse {
    string user_name = 1;
}

message BasicMixParamWithSerializerListResponse {
    repeated google.protobuf.Struct results = 1;
    int32 count = 2;
}

message BasicParamWithSerializerListRequest {
    repeated BasicParamWithSerializerRequest results = 1;
    int32 count = 2;
}

message BasicParamWithSerializerRequest {
    // @test=comment1
    // @test2=comment2
    string user_name = 1;
    // @test=test_in_serializer
    google.protobuf.Struct user_data = 2;
    string user_password = 3;
    bytes bytes_example = 4;
    repeated google.protobuf.Struct list_of_dict = 5;
}

message BasicProtoListChildListRequest {
    repeated BasicProtoListChildRequest results = 1;
    int32 count = 2;
}

message BasicProtoListChildListResponse {
    repeated BasicProtoListChildResponse results = 1;
    int32 count = 2;
}

message BasicProtoListChildRequest {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
}

message BasicProtoListChildResponse {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
}

message BasicServiceListResponse {
    repeated BasicServiceResponse results = 1;
    int32 count = 2;
}

message BasicServiceRequest {
    // @test=comment1
    // @test2=comment2
    string user_name = 1;
    // @test=test_in_serializer
    google.protobuf.Struct user_data = 2;
    string user_password = 3;
    bytes bytes_example = 4;
    repeated google.protobuf.Struct list_of_dict = 5;
}

message BasicServiceResponse {
    // @test=comment1
    // @test2=comment2
    string user_name = 1;
    // @test=test_in_serializer
    google.protobuf.Struct user_data = 2;
    bytes bytes_example = 3;
    repeated google.protobuf.Struct list_of_dict = 4;
}

message BasicTestNoMetaSerializerResponse {
    string value = 1;
}

message CustomMixParamForListRequest {
    repeated CustomMixParamForRequest results = 1;
    int32 count = 2;
}

message CustomMixParamForRequest {
    string user_name = 1;
}

message CustomNameForRequest {
    // @test=in_decorator
    string user_name = 1;
}

message CustomNameForResponse {
    string user_name = 1;
}

// Test comment for whole message
message CustomRetrieveResponseSpecialFieldsModelResponse {
    optional string uuid = 1;
    int32 default_method_field = 2;
    repeated google.protobuf.Struct custom_method_field = 3;
}

message DefaultValueDestroyRequest {
    int32 id = 1;
}

message DefaultValueListRequest {
}

message DefaultValueListResponse {
    repeated DefaultValueResponse results = 1;
    int32 count = 2;
}

message DefaultValuePartialUpdateRequest {
    optional int32 id = 1;
    optional string string_required_but_serializer_default = 2;
    optional int32 int_required_but_serializer_default = 3;
    optional bool boolean_required_but_serializer_default = 4;
    optional string string_default_but_serializer_default = 5;
    optional string string_nullable_default_but_serializer_default = 6;
    repeated string _partial_update_fields = 7;
    string string_required = 8;
    optional string string_blank = 9;
    optional string string_nullable = 10;
    optional string string_default = 11;
    optional string string_default_and_blank = 12;
    optional string string_null_default_and_blank = 13;
    int32 int_required = 14;
    optional int32 int_nullable = 15;
    optional int32 int_default = 16;
    bool boolean_required = 17;
    optional bool boolean_nullable = 18;
    optional bool boolean_default_false = 19;
    optional bool boolean_default_true = 20;
}

message DefaultValueRequest {
    optional int32 id = 1;
    optional string string_required_but_serializer_default = 2;
    optional int32 int_required_but_serializer_default = 3;
    optional bool boolean_required_but_serializer_default = 4;
    optional string string_default_but_serializer_default = 5;
    optional string string_nullable_default_but_serializer_default = 6;
    string string_required = 7;
    optional string string_blank = 8;
    optional string string_nullable = 9;
    optional string string_default = 10;
    optional string string_default_and_blank = 11;
    optional string string_null_default_and_blank = 12;
    int32 int_required = 13;
    optional int32 int_nullable = 14;
    optional int32 int_default = 15;
    bool boolean_required = 16;
    optional bool boolean_nullable = 17;
    optional bool boolean_default_false = 18;
    optional bool boolean_default_true = 19;
}

message DefaultValueResponse {
    optional int32 id = 1;
    optional string string_required_but_serializer_default = 2;
    optional int32 int_required_but_serializer_default = 3;
    optional bool boolean_required_but_serializer_default = 4;
    optional string string_default_but_serializer_default = 5;
    optional string string_nullable_default_but_serializer_default = 6;
    string string_required = 7;
    optional string string_blank = 8;
    optional string string_nullable = 9;
    optional string string_default = 10;
    optional string string_default_and_blank = 11;
    optional string string_null_default_and_blank = 12;
    int32 int_required = 13;
    optional int32 int_nullable = 14;
    optional int32 int_default = 15;
    bool boolean_required = 16;
    optional bool boolean_nullable = 17;
    optional bool boolean_default_false = 18;
    optional bool boolean_default_true = 19;
}

message DefaultValueRetrieveRequest {
    int32 id = 1;
}

message ExceptionStreamRaiseExceptionResponse {
    string id = 1;
}

message ForeignModelListRequest {
}

message ForeignModelListResponse {
    repeated ForeignModelResponse results = 1;
    int32 count = 2;
}

message ForeignModelResponse {
    optional string uuid = 1;
    string name = 2;
}

message ForeignModelRetrieveCustomResponse {
    string name = 1;
    string custom = 2;
}

message ForeignModelRetrieveCustomRetrieveRequest {
    string name = 1;
}

message ImportStructEvenInArrayModelRequest {
    optional string uuid = 1;
    repeated google.protobuf.Struct this_is_crazy = 2;
}

message ImportStructEvenInArrayModelResponse {
    optional string uuid = 1;
    repeated google.protobuf.Struct this_is_crazy = 2;
}

message ManyManyModelRequest {
    optional string uuid = 1;
    string name = 2;
    string test_write_only_on_nested = 3;
}

message ManyManyModelResponse {
    optional string uuid = 1;
    string name = 2;
}

message NoMetaRequest {
    string my_field = 1;
}

message RecursiveTestModelDestroyRequest {
    string uuid = 1;
}

message RecursiveTestModelListRequest {
}

message RecursiveTestModelListResponse {
    repeated RecursiveTestModelResponse results = 1;
    int32 count = 2;
}

message RecursiveTestModelPartialUpdateRequest {
    optional string uuid = 1;
    repeated string _partial_update_fields = 2;
    optional RecursiveTestModelRequest parent = 3;
    repeated RecursiveTestModelRequest children = 4;
}

message RecursiveTestModelRequest {
    optional string uuid = 1;
    optional RecursiveTestModelRequest parent = 2;
    repeated RecursiveTestModelRequest children = 3;
}

message RecursiveTestModelResponse {
    optional string uuid = 1;
    optional RecursiveTestModelResponse parent = 2;
    repeated RecursiveTestModelResponse children = 3;
}

message RecursiveTestModelRetrieveRequest {
    string uuid = 1;
}

message RelatedFieldModelDestroyRequest {
    string uuid = 1;
}

message RelatedFieldModelListRequest {
}

message RelatedFieldModelListResponse {
    repeated RelatedFieldModelResponse list_custom_field_name = 1;
    int32 count = 2;
}

message RelatedFieldModelPartialUpdateRequest {
    optional string uuid = 1;
    repeated ManyManyModelRequest many_many = 2;
    string custom_field_name = 3;
    repeated string _partial_update_fields = 4;
    repeated string many_many_foreigns = 5;
}

message RelatedFieldModelRequest {
    optional string uuid = 1;
    repeated ManyManyModelRequest many_many = 2;
    string custom_field_name = 3;
    repeated string many_many_foreigns = 4;
}

message RelatedFieldModelResponse {
    optional string uuid = 1;
    optional ForeignModelResponse foreign = 2;
    repeated ManyManyModelResponse many_many = 3;
    optional int32 slug_test_model = 4;
    repeated bool slug_reverse_test_model = 5;
    repeated string slug_many_many = 6;
    optional string proto_slug_related_field = 7;
    string custom_field_name = 8;
    repeated string many_many_foreigns = 9;
}

message RelatedFieldModelRetrieveRequest {
    string uuid = 1;
}

message SimpleRelatedFieldModelDestroyRequest {
    string uuid = 1;
}

message SimpleRelatedFieldModelListRequest {
}

message SimpleRelatedFieldModelListResponse {
    repeated SimpleRelatedFieldModelResponse results = 1;
    int32 count = 2;
}

message SimpleRelatedFieldModelPartialUpdateRequest {
    optional string uuid = 1;
    repeated string _partial_update_fields = 2;
    optional string foreign = 3;
    optional string slug_test_model = 4;
    repeated string many_many = 5;
    repeated string slug_many_many = 6;
    repeated string many_many_foreigns = 7;
}

message SimpleRelatedFieldModelRequest {
    optional string uuid = 1;
    optional string foreign = 2;
    optional string slug_test_model = 3;
    repeated string many_many = 4;
    repeated string slug_many_many = 5;
    repeated string many_many_foreigns = 6;
}

message SimpleRelatedFieldModelResponse {
    optional string uuid = 1;
    optional string foreign = 2;
    optional string slug_test_model = 3;
    repeated string many_many = 4;
    repeated string slug_many_many = 5;
    repeated string many_many_foreigns = 6;
}

message SimpleRelatedFieldModelRetrieveRequest {
    string uuid = 1;
}

message SpecialFieldsModelDestroyRequest {
    string uuid = 1;
}

message SpecialFieldsModelListRequest {
}

message SpecialFieldsModelListResponse {
    repeated SpecialFieldsModelResponse results = 1;
    int32 count = 2;
}

// Special Fields Model
// with two lines comment
message SpecialFieldsModelPartialUpdateRequest {
    optional string uuid = 1;
    repeated string _partial_update_fields = 2;
    optional google.protobuf.Struct meta_datas = 3;
    repeated int32 list_datas = 4;
}

// Special Fields Model
// with two lines comment
message SpecialFieldsModelRequest {
    optional string uuid = 1;
    optional google.protobuf.Struct meta_datas = 2;
    repeated int32 list_datas = 3;
}

// Special Fields Model
// with two lines comment
message SpecialFieldsModelResponse {
    optional string uuid = 1;
    optional google.protobuf.Struct meta_datas = 2;
    repeated int32 list_datas = 3;
    optional bytes binary = 4;
}

message SpecialFieldsModelRetrieveRequest {
    string uuid = 1;
}

message StreamInStreamInListResponse {
    repeated StreamInStreamInResponse results = 1;
    int32 count = 2;
}

message StreamInStreamInRequest {
    string name = 1;
}

message StreamInStreamInResponse {
    int32 count = 1;
}

message StreamInStreamToStreamRequest {
    string name = 1;
}

message StreamInStreamToStreamResponse {
    string name = 1;
}

message SyncUnitTestModelListWithExtraArgsRequest {
    bool archived = 1;
}

message UnitTestModelDestroyRequest {
    int32 id = 1;
}

message UnitTestModelListExtraArgsResponse {
    int32 count = 1;
    string query_fetched_datetime = 2;
    repeated UnitTestModelResponse results = 3;
}

message UnitTestModelListRequest {
}

message UnitTestModelListResponse {
    repeated UnitTestModelResponse results = 1;
    int32 count = 2;
}

message UnitTestModelListWithExtraArgsRequest {
    bool archived = 1;
}

message UnitTestModelPartialUpdateRequest {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
    repeated string _partial_update_fields = 4;
}

message UnitTestModelRequest {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
}

message UnitTestModelResponse {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
    int32 model_property = 4;
}

message UnitTestModelRetrieveRequest {
    int32 id = 1;
}

message UnitTestModelStreamRequest {
}

message UnitTestModelWithCacheDestroyRequest {
    int32 id = 1;
}

message UnitTestModelWithCacheInheritListWithStructFilterRequest {
    optional google.protobuf.Struct _filters = 1;
    optional google.protobuf.Struct _pagination = 2;
}

message UnitTestModelWithCacheListResponse {
    repeated UnitTestModelWithCacheResponse results = 1;
    int32 count = 2;
}

message UnitTestModelWithCacheListWithStructFilterRequest {
    optional google.protobuf.Struct _filters = 1;
    optional google.protobuf.Struct _pagination = 2;
}

message UnitTestModelWithCachePartialUpdateRequest {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
    repeated string _partial_update_fields = 4;
}

message UnitTestModelWithCacheRequest {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
}

message UnitTestModelWithCacheResponse {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
    int32 model_property = 4;
    string verify_custom_header = 5;
}

message UnitTestModelWithCacheRetrieveRequest {
    int32 id = 1;
}

message UnitTestModelWithCacheStreamRequest {
}

message UnitTestModelWithStructFilterDestroyRequest {
    int32 id = 1;
}

message UnitTestModelWithStructFilterEmptyWithFilterRequest {
    optional google.protobuf.Struct _filters = 1;
    optional google.protobuf.Struct _pagination = 2;
}

message UnitTestModelWithStructFilterListRequest {
    optional google.protobuf.Struct _filters = 1;
    optional google.protobuf.Struct _pagination = 2;
}

message UnitTestModelWithStructFilterListResponse {
    repeated UnitTestModelWithStructFilterResponse results = 1;
    int32 count = 2;
}

message UnitTestModelWithStructFilterPartialUpdateRequest {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
    repeated string _partial_update_fields = 4;
}

message UnitTestModelWithStructFilterRequest {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
}

message UnitTestModelWithStructFilterResponse {
    optional int32 id = 1;
    string title = 2;
    optional string text = 3;
    int32 model_property = 4;
}

message UnitTestModelWithStructFilterRetrieveRequest {
    int32 id = 1;
}

message UnitTestModelWithStructFilterStreamRequest {
}

